### PR TITLE
[SOAR-8977] Microsoft Log Analytics - Saved Search actions

### PR DESCRIPTION
**Pull request by igorski-r7**
### Description

Describe the proposed changes:

  - List Saved Searches
  - Get Saved Search
  - Create or Update Saved Search
  - Delete Saved Search

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [X] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [X] Screenshot of job output with the plugin changes
- [X] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [X] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [X] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [X] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [X] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [X] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [X] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [X] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [X] Work fully completed
- [X] Functional
  - [X] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [X] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [X] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [X] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [X] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
    "body": {
        "log": "Connect: Connecting...\nrapid7/Microsoft Log Analytics:1.2.0. Step name: list_all_searches\nUpdating auth token...\nAuthentication Token: ****************NH5Yg\nGetting shared key...\nShared Key: ****************BHQ==\nGetting workspace ID...\nWorkspace ID: ****************d2a2a\n",
        "meta": {},
        "output": {
            "saved_searches": [
                {
                    "id": "/subscriptions/0000-0000-000-00-0000/resourceGroups/test/providers/Microsoft.OperationalInsights/workspaces/test/savedSearches/test_create_saved_search",
                    "name": "test_create_saved_search",
                    "properties": {
                        "category": "Saved Search Test Category",
                        "displayName": "Create or Update Saved Search Test",
                        "query": "Test1Log1_CL"
                    }
}
            ]
        },
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/list_all_saved_searches.json
</summary>
</details>

<details>

```
{
    "body": {
        "log": "rapid7/Microsoft Log Analytics:1.2.0. Step name: get_saved_search\n",
        "meta": {},
        "output": {
            "id": "/subscriptions/0000-0000-000-00-0000/resourceGroups/test/providers/Microsoft.OperationalInsights/workspaces/test/savedSearches/test_create_saved_search",
            "name": "test_create_saved_search",
            "properties": {
                "category": "Saved Search Test Category",
                "displayName": "Create or Update Saved Search Test",
                "query": "Test1Log1_CL"
            }
        },
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/get_saved_search.json
</summary>
</details>

<details>

```
{
    "body": {
        "log": "rapid7/Microsoft Log Analytics:1.2.0. Step name: create_or_update_saved_search\n”,
        "meta": {},
        "output": {
            "id": "/subscriptions/0000-0000-000-00-0000/resourceGroups/test/providers/Microsoft.OperationalInsights/workspaces/test/savedSearches/test_create_saved_search",
            "name": "test_create_saved_search",
            "properties": {
                "category": "Saved Search Test Category",
                "displayName": "Create or Update Saved Search Test",
                "query": "Test1Log1_CL"
            }
        },
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/create_or_update_saved_search.json
</summary>
</details>

<details>

```
{
    "body": {
        "log": "rapid7/Microsoft Log Analytics:1.2.0. Step name: delete_saved_search\n",
        "meta": {},
        "output": {
            "deleted_saved_search": {
                "id": "/subscriptions/0000-0000-000-00-0000/resourceGroups/test/providers/Microsoft.OperationalInsights/workspaces/test/savedSearches/test_create_saved_search",
                "name": "test_create_saved_search",
                "properties": {
                    "category": "Saved Search Test Category",
                    "displayName": "Create or Update Saved Search Test",
                    "query": "Test1Log1_CL"
                }
            },
            "message": "Saved search test_create_saved_search has been deleted"
        },
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/delete_saved_search.json
</summary>
</details>

### Test

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Microsoft Log Analytics:1.2.0. Step name: list_all_saved_searches\nUpdating auth token...\nAuthentication Token: ****************sAekA\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/list_all_saved_searches.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Microsoft Log Analytics:1.2.0. Step name: get_saved_search\nUpdating auth token...\nAuthentication Token: ****************sAekA\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/get_saved_search.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Microsoft Log Analytics:1.2.0. Step name: create_or_update_saved_search\nUpdating auth token...\nAuthentication Token: ****************sAekA\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/create_or_update_saved_search.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Microsoft Log Analytics:1.2.0. Step name: delete_saved_search\nUpdating auth token...\nAuthentication Token: ****************sAekA\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.2.0 run < tests/delete_saved_search.json
</summary>
</details>

### UI

List All Saved Searches
List All Saved Searches

Get Saved Search
Get Saved Search

Create or Update Saved Search
Create or Update Saved Search

Delete Saved Search
Delete Saved Search

### CODE REVIEW

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/microsoft_log_analytics/help.md [hunk 1]
```diff
@@ -42,6 +46,232 @@ Example input:
...
+#### Delete Saved Search
+
+This action deletes the specified saved search in a given workspace.
+
+##### Input
+
+|Name|Type|Default|Required|Description|Enum|Example|
+|----|----|-------|--------|-----------|----|-------|
+|resource_group_name|string|None|True|Name of the resource group|None|ExampleResourceGroupName|
+|saved_search_name|string|None|True|Name of the saved search (case sensitive)|None|ExampleSavedSearchName|
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/help.md:58]:
```suggestion
|saved_search_name|string|None|True|Name of the saved search (case-sensitive)|None|ExampleSavedSearchName|
```

File: plugins/microsoft_log_analytics/help.md [hunk 2]
```diff
@@ -42,6 +46,232 @@ Example input:
...
+#### Get Saved Search
+
+This action gets the saved searche from Log Analytics by it's name.
+
+##### Input
+
+|Name|Type|Default|Required|Description|Enum|Example|
+|----|----|-------|--------|-----------|----|-------|
+|resource_group_name|string|None|True|Name of the resource group|None|ExampleResourceGroupName|
+|saved_search_name|string|None|True|Name of the saved search (case sensitive)|None|ExampleSavedSearchName|
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/help.md:106]:
```suggestion
|saved_search_name|string|None|True|Name of the saved search (case-sensitive)|None|ExampleSavedSearchName|
```

File: plugins/microsoft_log_analytics/help.md [hunk 3]
```diff
@@ -42,6 +46,232 @@ Example input:
...
+    "category": " Saved Search Test Category",
+    "displayName": "Get Saved Search Example",
+    "query": "* | where TimeGenerated > ago(1h)"
+  }
+}
+```
+
+#### List All Searches
+
+This action gets the saved searches for a given Log Analytics workspace.
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/help.md:207]:
```suggestion
This action retrieves all the saved searches for a given Log Analytics workspace.
```

File: plugins/microsoft_log_analytics/help.md [hunk 4]
```diff
@@ -271,13 +520,21 @@ Example output:
...
 
+#### tag
+
+|Name|Type|Required|Description|
+|----|----|--------|-----------|
+|Name|string|False|Tag's name|
+|Value|string|False|Tag's value|
+
 
 ## Troubleshooting
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/help.md:531]: Should we remove a section when it's empty?

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py [hunk 5]
```diff
@@ -13,7 +13,7 @@ def __init__(self):
         )
 
     def run(self, params={}):
-        interval = abs(params.get(Input.INTERVAL, 60))
+        interval = abs(params.get(Input.INTERVAL, 900))
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py:16]: Why this change?

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/api.py [hunk 6]
```diff
@@ -199,5 +206,109 @@ def send_log_data(
...
+        self._connection(
+            self.client_id,
+            self.client_secret,
+            self.tenant_id,
+            subscription_id,
+            resource_group_name,
+            workspace_name,
+            Endpoint.RESOURCE_MANAGEMENT,
+        )
+        api_version = "2020-08-01"
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/api.py:243]: Maybe api_version could be common to all the methods?

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py [hunk 7]
```diff
@@ -91,3 +93,67 @@ def return_non_empty_query_output(query_response_input: dict) -> dict:
     """
 
     return {"tables": list(filter(lambda x: x.get("rows"), query_response_input.get("tables")))}
+
+
+def remove_keys_from_saved_search(
+    sub_dictionaries: List[str], keys: List[str], input_list_of_dicts_or_dict: Union[List[dict], dict]
+) -> Union[List[dict], dict]:
+    """Method allows to remove version key from List ALl Saved Searches action's output.
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py:101]:
```suggestion
    """Method allows to remove version key from List All Saved Searches action's output.
```

**Review by jkaluza-rapid7 — CHANGES_REQUESTED** [review]

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py (hunk 5 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py:16]: According to the plugin.spec.yaml. I didn't notice that I missed that change that in previous PR 😞

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/endpoints.py [hunk 8]
```diff
@@ -14,3 +14,15 @@ class Endpoint:
...
+        + "/subscriptions/{}/resourcegroups/{}/providers/Microsoft.OperationalInsights/workspaces/{}/savedSearches/{}?api-version={}"
+    )
+    DELETE_SAVED_SEARCH = GET_SAVED_SEARCH
+    CREATE_OR_UPDATE_SAVED_SEARCH = GET_SAVED_SEARCH
+    LIST_ALL_SEARCHES = (
+        RESOURCE_MANAGEMENT
+        + "/subscriptions/{}/resourcegroups/{}/providers/Microsoft.OperationalInsights/workspaces/{}/savedSearches?api-version={}"
+    )
+
+    API_VERSION_SAVED_SEARCH = "2020-08-01"
```
Comment (mberezin-r7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/endpoints.py:28]: Is this indeed supposed to be a hardcoded date? Or was this for testing?

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py [hunk 9]
```diff
@@ -91,3 +93,67 @@ def return_non_empty_query_output(query_response_input: dict) -> dict:
...
+
+def return_non_empty(input_dict: Dict[str, Any]) -> Dict[Any, Any]:
+    """return_non_empty. Cleans up recusively the dictionary
+    :param input_dict:
+    :type input_dict: Dict[str, Any]
+    :rtype: Dict[Any, Any]
+    """
+    temp_dict = {}
+    for key, value in input_dict.items():
+        if value is not None and value != "" and value != []:
```
Comment (mberezin-r7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py:152]: I think this is exactly what `if not value` resolves to. We could replace this with:
```suggestion
        if not value:
```
However, I also appreciate not making any assumptions, and being very exact in your definition. We may, however, need to check `if not value:` Python definition to make sure this covers every edge case.

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py [hunk 10]
```diff
@@ -91,3 +93,67 @@ def return_non_empty_query_output(query_response_input: dict) -> dict:
...
+    :rtype: List[dict]
+    """
+
+    for element in input_saved_search_list:
+        if "name" not in element:
+            element["name"] = element.get("id").split("/")[-1]
+    return input_saved_search_list
+
+
+def return_non_empty(input_dict: Dict[str, Any]) -> Dict[Any, Any]:
```
Comment (mberezin-r7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py:144]: Is it desired that this method can only accept an input_dict where the keys are strings? Can we not handle dicts where the keys are ints?

File: plugins/microsoft_log_analytics/unit_test/test_create_or_update_saved_search.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        self.assertEqual(expected_response, response)
+
+    @parameterized.expand(
+        [
+            (mock_request_201_invalid_json, PluginException.causes[PluginException.Preset.INVALID_JSON]),
+            (mock_request_400, Message.BAD_REQUEST_MESSAGE),
+            (mock_request_403, PluginException.causes[PluginException.Preset.UNAUTHORIZED]),
+            (mock_request_409, Message.CONFLICTED_STATE_OF_OBJECT_MESSAGE),
+            (mock_request_429, PluginException.causes[PluginException.Preset.RATE_LIMIT]),
+            (mock_request_500, PluginException.causes[PluginException.Preset.UNKNOWN]),
```
Comment (mberezin-r7) [inline plugins/microsoft_log_analytics/unit_test/test_create_or_update_saved_search.py:82]: I think we can return 500 errors as a SERVER_ERROR or SERVICE_UNAVAILABLE preset: https://github.com/rapid7/komand-plugin-sdk-python/blob/master/insightconnect_plugin_runtime/exceptions.py#L69

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/endpoints.py (hunk 8 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/endpoints.py:28]: It is hardcoded in order for easy change in the future (if needed)

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py (hunk 9 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py:152]: By means of `if not value:` that will miss all the boolean values with `False`

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py (hunk 10 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/icon_microsoft_log_analytics/util/tools.py:144]: in that case every single JSON key is `str`

File: plugins/microsoft_log_analytics/plugin.spec.yaml [hunk 12]
```diff
@@ -147,6 +200,167 @@ actions:
...
+        description: Name of the saved search (case sensitive)
+        type: string
+        required: true
+        example: ExampleSavedSearchName
+      properties:
+        title: Properties
+        description: Saved search properties object
+        type: properties
+        required: true
+        example: '{"properties":{"category":"Saved Search Test Category","displayName":"Create or Update Saved Search Test","functionAlias":"heartbeat_func","functionParameters":"a:int=1","query":"Heartbeat | summarize Count() by Computer | take a","tags":[{"name":"Group","value":"Computer"}]}}'
```
[the remark this answers is not in the document]
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/plugin.spec.yaml:236]: I don't know better way to put an example there. It's full object but separated at the UI site - so that should be (imo) readable for customer.

**Review by jkaluza-rapid7 — APPROVED** [review]

File: plugins/microsoft_log_analytics/help.md [hunk 13]
```diff
@@ -271,13 +520,21 @@ Example output:
...
 
 _This plugin does not contain any troubleshooting information._
 
 # Version History
 
+* 1.2.0 - Added new actions Create or Update Saved Search, Get Saved Search, Delete Saved Search, and List All Saved Searches
 * 1.1.0 - Search trigger thats run Log Analytics query every interval time
```
Comment (mberezin-r7) [inline plugins/microsoft_log_analytics/help.md:538]:
```suggestion
* 1.2.0 - Add new actions Create or Update Saved Search, Get Saved Search, Delete Saved Search, and List All Saved Searches
* 1.1.0 - Search trigger that run Log Analytics query every interval time
```

**Review by mberezin-r7 — APPROVED** [review]
One minor grammar comment, otherwise looks good to me!

File: plugins/microsoft_log_analytics/help.md (hunk 13 above)
Comment (blaxminarayan-r7) [reply plugins/microsoft_log_analytics/help.md:538]: Updated the change for changelog 1.2.0. Not changing the description for 1.1.0 as it is already released.
